hammerblade/torch/kernel/brg_element_for.hpp

```diff
@@ -0,0 +1,356 @@
+    uint32_t strides;
+#ifdef HB_EMUL
+    uint64_t data;
+#else
+    uint32_t data;
+#endif
+    uint32_t cur_loc;
+
+  public:
+    BRGIteratorTensor(bsg_tensor_t* t, uint32_t start_loc = 0) :
+      N(t->N),
+      dims(t->dims),
+      strides(*(uint32_t*)((intptr_t) t->strides)),
+      data((intptr_t) t->data),
+      cur_loc(start_loc) {
+        bsg_assert(cur_loc < N);
+        data += start_loc * strides;
+      }
+
+    uint32_t numel() {
+      return N;
+    }
+
+    //-----------------
+    // post increment
+    //-----------------
+    BRGIteratorTensor& operator ++ (int) {
+      data += strides;
+      cur_loc++;
+      return *this;
+    }
+
+    //-----------------------------------
+    // *accessor, just like c++ iterators
+    //-----------------------------------
+    T operator*() {
+      bsg_assert(cur_loc < N);
+      return (T)(intptr_t)data;
+    }
+
+};
+
+// =========================================================
+// Helper function to get the argument type of lambda function
+//
+// Creazy meta-programming in here ...
+// Please refer to https://stackoverflow.com/questions/28105371/
+// is-it-possible-to-figure-out-the-parameter-type-and-return-type-of-a-polymorphic
+// and
+// http://coliru.stacked-crooked.com/a/e19151a5c245d7c3
+//==========================================================
+
+namespace function_traits {
+  //--------------------------
+  // for lambda and functors
+  //--------------------------
+  template <typename T>
+  struct traits : public traits<decltype(&T::operator())>{};
+
+  template <typename ClassType, typename ReturnType, typename... Args>
+  struct traits<ReturnType(ClassType::*)(Args...) const> {
+    using result_type = ReturnType;
+    template <size_t i> struct arg {
+        using type = typename std::tuple_element<i, std::tuple<Args...>>::type;
+    };
+  };
+
+  //--------------------------
+  // for function pointers
+  //--------------------------
+  template <typename ReturnType, typename... Args>
+  struct traits<ReturnType(*)(Args...)> : public traits<ReturnType(Args...)>{};
+
+  template <typename ReturnType, typename... Args>
+  struct traits<ReturnType(Args...)> {
+    using result_type = ReturnType;
+    template <size_t i> struct arg {
+        using type = typename std::tuple_element<i, std::tuple<Args...>>::type;
+    };
+  };
+}
+
+// =========================================================
+// Element-wise for -- Binary ops
+//
+// This function iterates over all elements, starting from element 0
+//
+// We would like the synatx to be like this:
+// __attribute__ ((noinline))  int tensorlib_add(bsg_tensor_t* c_p,
+//                                               bsg_tensor_t* a_p,
+//                                               bsg_tensor_t* b_p,
+//                                               float* alpha) {
+//    brg_elementwise_for(t0_p, t1_p, t2_p,
+//        [&](float a, float b) {
+//          return a + alpha * b;
+//        });
```

Comment:
Thought we decided to hold-off lambda functions for a while... jk, I love this. This helps us decouple optimizing parallel primitives vs optimizing actual computation.

Reply:
oh well you know ... things happen ...

So the cool part here is that we figure out the data type by looking at the lambda function ;)